File: day5/go/part_1.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type Rule [2]int
type Update []int

func parseInput(filePath string) ([]Rule, []Update) {
	file, err := os.Open(filePath)
	if err != nil {
		panic(err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	var rules []Rule
	var updates []Update
	section := 0

	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" {
			section++
			continue
		}

		if section == 0 {
			parts := strings.Split(line, "|")
			a, _ := strconv.Atoi(parts[0])
			b, _ := strconv.Atoi(parts[1])
			rules = append(rules, Rule{a, b})
		} else {
			var update Update
			for _, numStr := range strings.Split(line, ",") {
				num, _ := strconv.Atoi(numStr)
				update = append(update, num)
			}
			updates = append(updates, update)
		}
	}

	return rules, updates
}

func isUpdateOrdered(update Update, rules []Rule) bool {
	pageIndices := make(map[int]int)
	for i, page := range update {
		pageIndices[page] = i
	}

	for _, rule := range rules {
		if i1, ok1 := pageIndices[rule[0]]; ok1 {
			if i2, ok2 := pageIndices[rule[1]]; ok2 {
				if i1 > i2 {
					return false
				}
			}
		}
	}
	return true
}

func main() {
	rules, updates := parseInput("input.txt")
	middleSum := 0

	for _, update := range updates {
		if isUpdateOrdered(update, rules) {
			middleSum += update[len(update)/2]
		}
	}

	fmt.Println(middleSum)
}

File: day5/java/Part1.java
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.*;

public class Part1 {
    private static class Rule {
        int a, b;
        Rule(int a, int b) {
            this.a = a;
            this.b = b;
        }
    }

    private static class Input {
        List<Rule> rules;
        List<List<Integer>> updates;

        Input(List<Rule> rules, List<List<Integer>> updates) {
            this.rules = rules;
            this.updates = updates;
        }
    }

    private static Input parseInput(String filePath) throws Exception {
        List<String> lines = Files.readAllLines(Paths.get(filePath));
        List<Rule> rules = new ArrayList<>();
        List<List<Integer>> updates = new ArrayList<>();
        int section = 0;

        for (String line : lines) {
            line = line.trim();
            if (line.isEmpty()) {
                section++;
                continue;
            }

            if (section == 0) {
                String[] parts = line.split("\\|");
                rules.add(new Rule(
                    Integer.parseInt(parts[0]),
                    Integer.parseInt(parts[1])
                ));
            } else {
                List<Integer> update = new ArrayList<>();
                for (String num : line.split(",")) {
                    update.add(Integer.parseInt(num));
                }
                updates.add(update);
            }
        }

        return new Input(rules, updates);
    }

    private static boolean isUpdateOrdered(List<Integer> update, List<Rule> rules) {
        Map<Integer, Integer> pageIndices = new HashMap<>();
        for (int i = 0; i < update.size(); i++) {
            pageIndices.put(update.get(i), i);
        }

        for (Rule rule : rules) {
            if (pageIndices.containsKey(rule.a) && pageIndices.containsKey(rule.b)) {
                if (pageIndices.get(rule.a) > pageIndices.get(rule.b)) {
                    return false;
                }
            }
        }
        return true;
    }

    public static void main(String[] args) throws Exception {
        Input input = parseInput("input.txt");
        int middleSum = input.updates.stream()
            .filter(update -> isUpdateOrdered(update, input.rules))
            .mapToInt(update -> update.get(update.size() / 2))
            .sum();

        System.out.println(middleSum);
    }
}
